File: python/python-sdk/kestrapy/api/metrics_api.py
from typing import Any, Dict, List, Optional

from kestrapy.base_api import BaseApi


class MetricsApi(BaseApi):
    """Metrics (`/api/v1/{tenant}/metrics`).

    Task/flow metric lookups. The EE controller is gated by the ``EXECUTION``
    resource, so routes may answer 403 when the feature is unavailable.
    """

    def execution_metrics(
        self,
        execution_id: str,
        tenant: str,
        page: Optional[int] = None,
        size: Optional[int] = None,
        sort: Optional[List[str]] = None,
        task_run_id: Optional[str] = None,
        task_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        """A page of the metrics emitted by one execution, optionally narrowed to
        a task run or task id. GET /api/v1/{tenant}/metrics/{executionId}."""
        path = self._tenant_path(tenant, "metrics", execution_id)
        params = list(
            self._build_query_params(
                page=page, size=size, taskRunId=task_run_id, taskId=task_id
            ).items()
        )
        self._append_repeated_param(params, "sort", sort)
        return self._raw_json_request("GET", path, params=params)

    def flow_metric_names(self, namespace: str, flow_id: str, tenant: str) -> List[str]:
        """Distinct metric names emitted by a flow.
        GET /api/v1/{tenant}/metrics/names/{namespace}/{flowId}."""
        path = self._tenant_path(tenant, "metrics", "names", namespace, flow_id)
        return self._json_list_request("GET", path, str)

    def task_metric_names(
        self, namespace: str, flow_id: str, task_id: str, tenant: str
    ) -> List[str]:
        """Distinct metric names emitted by one task of a flow.
        GET /api/v1/{tenant}/metrics/names/{namespace}/{flowId}/{taskId}."""
        path = self._tenant_path(tenant, "metrics", "names", namespace, flow_id, task_id)
        return self._json_list_request("GET", path, str)

    def tasks_with_metrics(
        self, namespace: str, flow_id: str, tenant: str
    ) -> List[str]:
        """Task ids of a flow that have emitted metrics.
        GET /api/v1/{tenant}/metrics/tasks/{namespace}/{flowId}."""
        path = self._tenant_path(tenant, "metrics", "tasks", namespace, flow_id)
        return self._json_list_request("GET", path, str)

    def aggregate_flow_metric(
        self,
        namespace: str,
        flow_id: str,
        metric: str,
        tenant: str,
        start_date: Optional[str] = None,
        end_date: Optional[str] = None,
        aggregation: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Time-bucketed aggregation of one metric across a flow's executions.
        GET /api/v1/{tenant}/metrics/aggregates/{namespace}/{flowId}/{metric}."""
        path = self._tenant_path(tenant, "metrics", "aggregates", namespace, flow_id, metric)
        params = self._build_query_params(
            startDate=start_date, endDate=end_date, aggregation=aggregation
        )
        return self._raw_json_request("GET", path, params=params)

    def aggregate_task_metric(
        self,
        namespace: str,
        flow_id: str,
        task_id: str,
        metric: str,
        tenant: str,
        start_date: Optional[str] = None,
        end_date: Optional[str] = None,
        aggregation: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Time-bucketed aggregation of one metric for a single task across a
        flow's executions.
        GET /api/v1/{tenant}/metrics/aggregates/{namespace}/{flowId}/{taskId}/{metric}."""
        path = self._tenant_path(
            tenant, "metrics", "aggregates", namespace, flow_id, task_id, metric
        )
        params = self._build_query_params(
            startDate=start_date, endDate=end_date, aggregation=aggregation
        )
        return self._raw_json_request("GET", path, params=params)

File: go-sdk/kestra_api_client/metrics_api.go
package kestra_api_client

import "context"

// MetricsAPI covers /api/v1/{tenant}/metrics — task/flow metric lookups. The EE
// controller is gated by the EXECUTION resource.
type MetricsAPI struct {
	baseAPI
}

// ExecutionMetrics returns a page of the metrics emitted by one execution,
// optionally narrowed to a task run or task id. Backs GET
// /api/v1/{tenant}/metrics/{executionId}.
func (a *MetricsAPI) ExecutionMetrics(ctx context.Context, executionId, tenant string, page, size *int, sort []string, taskRunId, taskId *string) (map[string]interface{}, error) {
	params := buildQueryParams("page", page, "size", size, "taskRunId", taskRunId, "taskId", taskId)
	appendRepeatedParam(params, "sort", sort)
	return doJSON[map[string]interface{}](&a.baseAPI, ctx, "GET", tenantPath(tenant, "metrics", executionId), nil, params)
}

// FlowMetricNames lists the distinct metric names emitted by a flow. Backs GET
// /api/v1/{tenant}/metrics/names/{namespace}/{flowId}.
func (a *MetricsAPI) FlowMetricNames(ctx context.Context, namespace, flowId, tenant string) ([]string, error) {
	return doJSON[[]string](&a.baseAPI, ctx, "GET", tenantPath(tenant, "metrics", "names", namespace, flowId), nil, nil)
}

// TaskMetricNames lists the distinct metric names emitted by one task of a flow.
// Backs GET /api/v1/{tenant}/metrics/names/{namespace}/{flowId}/{taskId}.
func (a *MetricsAPI) TaskMetricNames(ctx context.Context, namespace, flowId, taskId, tenant string) ([]string, error) {
	return doJSON[[]string](&a.baseAPI, ctx, "GET", tenantPath(tenant, "metrics", "names", namespace, flowId, taskId), nil, nil)
}

// TasksWithMetrics lists the task ids of a flow that have emitted metrics. Backs
// GET /api/v1/{tenant}/metrics/tasks/{namespace}/{flowId}.
func (a *MetricsAPI) TasksWithMetrics(ctx context.Context, namespace, flowId, tenant string) ([]string, error) {
	return doJSON[[]string](&a.baseAPI, ctx, "GET", tenantPath(tenant, "metrics", "tasks", namespace, flowId), nil, nil)
}

// AggregateFlowMetric returns the time-bucketed aggregation of one metric across
// a flow's executions. `aggregation` is one of avg/sum/min/max (default sum);
// startDate/endDate are ISO-8601 timestamps (optional). Backs GET
// /api/v1/{tenant}/metrics/aggregates/{namespace}/{flowId}/{metric}.
func (a *MetricsAPI) AggregateFlowMetric(ctx context.Context, namespace, flowId, metric, tenant string, startDate, endDate, aggregation *string) (map[string]interface{}, error) {
	params := buildQueryParams("startDate", startDate, "endDate", endDate, "aggregation", aggregation)
	return doJSON[map[string]interface{}](&a.baseAPI, ctx, "GET", tenantPath(tenant, "metrics", "aggregates", namespace, flowId, metric), nil, params)
}

// AggregateTaskMetric returns the time-bucketed aggregation of one metric for a
// single task across a flow's executions. Backs GET
// /api/v1/{tenant}/metrics/aggregates/{namespace}/{flowId}/{taskId}/{metric}.
func (a *MetricsAPI) AggregateTaskMetric(ctx context.Context, namespace, flowId, taskId, metric, tenant string, startDate, endDate, aggregation *string) (map[string]interface{}, error) {
	params := buildQueryParams("startDate", startDate, "endDate", endDate, "aggregation", aggregation)
	return doJSON[map[string]interface{}](&a.baseAPI, ctx, "GET", tenantPath(tenant, "metrics", "aggregates", namespace, flowId, taskId, metric), nil, params)
}
